lib/createReconciler.lua

```diff
@@ -0,0 +1,294 @@
+			local newElement = ChildUtils.getChildByKey(newChildElements, childKey)
+			local newNode = updateVirtualNode(childNode, newElement)
+
+			if newNode ~= nil then
+				virtualNode.children[childKey] = newNode
+			else
+				removeKeys[childKey] = true
+			end
+		end
+
+		for childKey in pairs(removeKeys) do
+			virtualNode.children[childKey] = nil
+		end
+
+		-- Added children
+		for childKey, newElement in ChildUtils.iterateChildren(newChildElements) do
+			local childNode = virtualNode.children[childKey]
+
+			local concreteKey = childKey
+			if childKey == ChildUtils.UseParentKey then
+				concreteKey = virtualNode.key
+			end
+
+			if childNode == nil then
+				virtualNode.children[childKey] = mountVirtualNode(newElement, virtualNode.hostObject, concreteKey)
+			end
+		end
+	end
+
+	--[[
+		Unmounts the given virtual node and releases any held resources.
+	]]
+	local function unmountVirtualNode(virtualNode)
+		assert(Type.of(virtualNode) == Type.VirtualNode)
+
+		local kind = ElementKind.of(virtualNode.currentElement)
+
+		if kind == ElementKind.Host then
+			renderer.unmountHostNode(reconciler, virtualNode)
+		elseif kind == ElementKind.Function then
+			for _, child in pairs(virtualNode.children) do
+				unmountVirtualNode(child)
+			end
+		elseif kind == ElementKind.Stateful then
+			virtualNode.instance:__unmount()
+		elseif kind == ElementKind.Portal then
+			error("NYI")
+		else
+			error(("Unknown ElementKind %q"):format(tostring(kind), 2))
+		end
+	end
+
+	local function updateFunctionVirtualNode(virtualNode, newElement)
+		local renderResult = newElement.component(newElement.props)
+
+		updateVirtualNodeChildren(virtualNode, renderResult)
+	end
+
+	--[[
+		Update the given virtual node using a new element describing what it
+		should transform into.
+
+		`updateVirtualNode` will return a new virtual node that should replace the
+		passed in virtual node. This is because a virtual node can be updated
+		with an element referencing a different component!
+
+		In that case, `updateVirtualNode` will unmount the input virtual node,
+		mount a new virtual node, and return it in this case, while also issuing
+		a warning to the user.
+	]]
+	function updateVirtualNode(virtualNode, newElement)
+		assert(Type.of(virtualNode) == Type.VirtualNode)
+		assert(Type.of(newElement) == Type.Element or typeof(newElement) == "boolean" or newElement == nil)
+
+		if typeof(newElement) == "boolean" or newElement == nil then
+			unmountVirtualNode(virtualNode)
+			return nil
+		end
+
+		if virtualNode.currentElement.component ~= newElement.component then
+			-- TODO: Better message
+			warn("Component changed type!")
+
+			local hostParent = virtualNode.hostParent
+			local key = virtualNode.key
+
+			unmountVirtualNode(virtualNode)
+			return mountVirtualNode(newElement, hostParent, key)
+		end
+
+		local kind = ElementKind.of(newElement)
+
+		if kind == ElementKind.Host then
+			return renderer.updateHostNode(reconciler, virtualNode, newElement)
```

Comment:
I just discovered something: at this point, nobody actually updates `node.currentElement`.
This means that rendering element A, update with element B, and then update again with element C it will compare against element A instead of element B, which could result in inaccuracies (e.g. set property in A, change it in B, attempt to revert in C). Presumably the reconciler should be responsible for setting `node.currentElement = newElement`